src/easylink/pipeline_graph.py

```diff
@@ -21,9 +21,91 @@ class PipelineGraph(ImplementationGraph):
 
     def __init__(self, config: Config) -> None:
         super().__init__(incoming_graph_data=config.schema.get_implementation_graph())
+        self.merge_combined_implementations(config)
         self.update_slot_filepaths(config)
         self = nx.freeze(self)
 
+    def merge_combined_implementations(self, config):
+        for (
+            combined_implementation,
+            combined_implementation_config,
+        ) in config.pipeline.combined_implementations.items():
+
+            # Find all nodes with the same implementation name
+            nodes_to_merge = [
+                node
+                for node, data in self.nodes(data=True)
+                if data["implementation"].combined_name == combined_implementation
+            ]
+            if not nodes_to_merge:
+                continue
+
+            implemented_steps = [
+                step
+                for node in nodes_to_merge
+                for step in self.nodes[node]["implementation"].schema_steps
+            ]
+            metadata_steps = self.nodes[nodes_to_merge[0]]["implementation"].metadata_steps
+            self.validate_implementation_topology(nodes_to_merge, metadata_steps)
+            input_slots = []
+            output_slots = []
+            in_edge_params = []
+            combined_edges = []
+
+            for node in nodes_to_merge:
+                for pred, succ, data in self.in_edges(node, data=True):
+                    if pred not in nodes_to_merge:
+                        input_slots.append(data["input_slot"])
+                        in_edge_params.append(EdgeParams.from_graph_edge(pred, succ, data))
+                        combined_edges.append(
+                            EdgeParams.from_graph_edge(pred, combined_implementation, data)
+                        )
+
+                for _, succ, data in self.out_edges(node, data=True):
+                    if succ not in nodes_to_merge:
+                        output_slots.append(data["output_slot"])
+                        combined_edges.append(
+                            EdgeParams.from_graph_edge(combined_implementation, succ, data)
+                        )
+            self.validate_edge_params(in_edge_params)
+
+            for slots in (input_slots, output_slots):
+                seen_slots = []
+                seen_names = []
+                seen_env_vars = []
+                for slot in slots:
+                    # Check for duplicate names
+                    if slot.name in seen_names and slot not in seen_slots:
```

Comment:
I'm thinking of something like the following situation, where we may have duplicates for more or less legitimate reasons:

```  
step_1:
    parallel:
      - input_data_file: input_file_1
        implementation:
            name: step_1_python_pandas
      - input_data_file: input_file_2
        combined_implementation_key: steps_1_and_2_combined
      - input_data_file: input_file_3
        combined_implementation_key: steps_1_and_2_combined
  step_2:
    combined_implementation_key: steps_1_and_2_combined
```
Where you're trying to combine slots that are duplicate, but that's OK because they're literally the same slot, so they show up in the list of input slots twice. This might be kind of an artificial example; I'm not sure how realistic it is to expect this to come up in the wild.

This fails our current setup anyway because both step_1 and step_2 take in DUMMY_CONTAINER_MAIN_<...> from different sources, of course.

If you agree this is something we should watch out for, i can add a comment in code as well.

Reply:
I don't think that one should be valid. Kind of by definition, those environment variables are going to collide, and I think that is "right" -- this is not a permissible combined implementation.

Is there a problem I'm not foreseeing with removing the `seen_slots` check altogether?

Reply:
fine by me! If we find some valid example where it breaks it will throw an error and we can figure out how to handle it then.